README.md
<h1 align="center">IMGTAGGEN</h1>

<p align="center">
<a href="https://imgtaggen.codewithshin.com/" rel="nofollow">Image Tag Generator</a>
</p>

<p align="center">
<a href="https://github.com/sponsors/shinokada" target="_blank"><img src="https://img.shields.io/static/v1?label=Sponsor&message=%E2%9D%A4&logo=GitHub&color=%23fe8e86" alt="Sponsor" height="22" width="102"></a>
<a href="https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps"><img src="https://img.shields.io/badge/PWA-enabled-brightgreen" alt="PWA Shield"></a>
<a href="https://www.npmjs.com/package/imgtaggen" rel="nofollow" target="_blank"><img src="https://img.shields.io/npm/v/imgtaggen" alt="npm" height="22" width="97"></a>
<a href="https://twitter.com/shinokada" rel="nofollow" target="_blank"><img src="https://img.shields.io/badge/created%20by-@shinokada-4BBAAB.svg" alt="Created by Shin Okada" height="22" width="161"></a>
<a href="https://opensource.org/licenses/MIT" rel="nofollow" target="_blank"><img src="https://img.shields.io/github/license/shinokada/imgtaggen" alt="License" height="22" width="86"></a>
<a href="https://www.npmjs.com/package/imgtaggen" rel="nofollow" target="_blank"><img src="https://img.shields.io/npm/dw/imgtaggen.svg" alt="npm" height="22" width="152"></a>
</p>

<p align="center">
<picture>
  <img
    src="public/images/imgtaggen-800-optimized.png"
    style="width: 100%; aspect-ratio: 1.791044776119403"
    decoding="async"
    alt="Image tag generator"
  />
</picture>
</p>


A CLI tool for generating a responsive image tag with support for AVIF and WebP formats. It will also calculate image ratio.

It is recommended to use this script in conjunction with [bimgc](https://www.npmjs.com/package/bimgc).

## Installation

```
npm i -g imgtaggen
imgtaggen --version
cd path/to/your/project
```

## Options

imgtaggen [args]

|Options|Description|Default|
|--|--|--|
|--noavif|Do not generate an AVIF image tag |[boolean] [default: false]|
|--nowebp|Do not generate a WEBP image tag |[boolean] [default: false]|
|--noclipboard|Do not copy the image tag to the clipboard|[boolean] [default: false]|
|--version|Show version number ||
|--help|Show help||


## Usages

Add an image path with `imgtaggen` command.

```bash
imgtaggen public/images/imgtaggen.png
```

This will copy the following to your clipboard.

```html
<picture>
  <source
    type="image/avif"
    srcset="public/images/imgtaggen-100.avif?width=100 100w, public/images/imgtaggen-200.avif?width=200 200w, public/images/imgtaggen-400.avif?width=400 400w, public/images/imgtaggen-800.avif?width=800 800w" />
  <source
    type="image/webp"
    srcset="public/images/imgtaggen-100.webp?width=100 100w, public/images/imgtaggen-200.webp?width=200 200w, public/images/imgtaggen-400.webp?width=400 400w, public/images/imgtaggen-800.webp?width=800 800w" />
  <img
    src="public/images/imgtaggen.png"
    srcset="public/images/imgtaggen-100.png?width=100 100w, public/images/imgtaggen-200.png?width=200 200w, public/images/imgtaggen-400.png?width=400 400w, public/images/imgtaggen-800.png?width=800 800w"
    sizes="(max-width: 800px) 100vw, 50vw"
    style="width: 100%; aspect-ratio: 1.791044776119403"
    loading="lazy"
    decoding="async"
    alt="My awesome image"
  />
</picture>
```

You just need to paster it and modify the `alt` text and `sizes` values to match your requirements.

## Sizes

The default outputs are 100, 200, 400, and 800. To specify a different set of outputs, use the -s or --sizes option.

```bash
imgtaggen public/images/imgtaggen.png -s 100 200
```

Output:

```html
<picture>
  <source
    type="image/avif"
    srcset="public/images/imgtaggen-100.avif?width=100 100w, public/images/imgtaggen-200.avif?width=200 200w" />
  <source
    type="image/webp"
    srcset="public/images/imgtaggen-100.webp?width=100 100w, public/images/imgtaggen-200.webp?width=200 200w" />
  <img
    src="public/images/imgtaggen.png"
    srcset="public/images/imgtaggen-100.png?width=100 100w, public/images/imgtaggen-200.png?width=200 200w"
    sizes="(max-width: 800px) 100vw, 50vw"
    style="width: 100%; aspect-ratio: 1.791044776119403"
    loading="lazy"
    decoding="async"
    alt="My awesome image"
  />
</picture>
```

## alt option

The default alt text is "My awesome image", and you can use the `-a` or
`--alt` option to specify a different alt text.

```bash
imgtaggen public/images/imgtaggen.png -a 'My alt text'
```

Output:

```html
<picture>
  <source
    type="image/avif"
    srcset="public/images/imgtaggen-100.avif?width=100 100w, public/images/imgtaggen-200.avif?width=200 200w, public/images/imgtaggen-400.avif?width=400 400w, public/images/imgtaggen-800.avif?width=800 800w" />
  <source
    type="image/webp"
    srcset="public/images/imgtaggen-100.webp?width=100 100w, public/images/imgtaggen-200.webp?width=200 200w, public/images/imgtaggen-400.webp?width=400 400w, public/images/imgtaggen-800.webp?width=800 800w" />
  <img
    src="public/images/imgtaggen.png"
    srcset="public/images/imgtaggen-100.png?width=100 100w, public/images/imgtaggen-200.png?width=200 200w, public/images/imgtaggen-400.png?width=400 400w, public/images/imgtaggen-800.png?width=800 800w"
    sizes="(max-width: 800px) 100vw, 50vw"
    style="width: 100%; aspect-ratio: 1.791044776119403"
    loading="lazy"
    decoding="async"
    alt="My alt text"
  />
</picture>
```

## No avif block

The option `--noavif` will not output avif block.

```bash
imgtaggen public/images/imgtaggen.png --noavif
```

Output:

```html
<picture>
  <source
    type="image/webp"
    srcset="public/images/imgtaggen-100.webp?width=100 100w, public/images/imgtaggen-200.webp?width=200 200w, public/images/imgtaggen-400.webp?width=400 400w, public/images/imgtaggen-800.webp?width=800 800w" />
  <img
    src="public/images/imgtaggen.png"
    srcset="public/images/imgtaggen-100.png?width=100 100w, public/images/imgtaggen-200.png?width=200 200w, public/images/imgtaggen-400.png?width=400 400w, public/images/imgtaggen-800.png?width=800 800w"
    sizes="(max-width: 800px) 100vw, 50vw"
    style="width: 100%; aspect-ratio: 1.791044776119403"
    loading="lazy"
    decoding="async"
    alt="My awesome image"
  />
</picture>
```

## No webp block

The option `--nowebp` will not output avif block.

```bash
imgtaggen public/images/imgtaggen.png --nowebp
```

Output:

```html
<picture>
  <source
    type="image/avif"
    srcset="public/images/imgtaggen-100.avif?width=100 100w, public/images/imgtaggen-200.avif?width=200 200w, public/images/imgtaggen-400.avif?width=400 400w, public/images/imgtaggen-800.avif?width=800 800w" />
  <img
    src="public/images/imgtaggen.png"
    srcset="public/images/imgtaggen-100.png?width=100 100w, public/images/imgtaggen-200.png?width=200 200w, public/images/imgtaggen-400.png?width=400 400w, public/images/imgtaggen-800.png?width=800 800w"
    sizes="(max-width: 800px) 100vw, 50vw"
    style="width: 100%; aspect-ratio: 1.791044776119403"
    loading="lazy"
    decoding="async"
    alt="My awesome image"
  />
</picture>
```

## No webp & no avif

```bash
imgtaggen public/images/imgtaggen.png --nowebp --noavif
```

Output:

```html
<picture>
  <img
    src="public/images/imgtaggen.png"
    srcset="public/images/imgtaggen-100.png?width=100 100w, public/images/imgtaggen-200.png?width=200 200w, public/images/imgtaggen-400.png?width=400 400w, public/images/imgtaggen-800.png?width=800 800w"
    sizes="(max-width: 800px) 100vw, 50vw"
    style="width: 100%; aspect-ratio: 1.791044776119403"
    loading="lazy"
    decoding="async"
    alt="My awesome image"
  />
</picture>
```

## No clipboard

The option`--noclipboard` will display the output in the terminal instead of copying it to the clipboard.

```bash
imgtaggen public/images/imgtaggen.png --noclipboard
```

This command outputs the following on your terminal.

```html
Generating image tag...
noavif: false
nowebp: false
noclipboard: true
sizes: [ 100, 200, 400, 800 ]
input file: public/images/imgtaggen.png
<picture>
  <source
    type="image/avif"
    srcset="public/images/imgtaggen-100.avif?width=100 100w, public/images/imgtaggen-200.avif?width=200 200w, public/images/imgtaggen-400.avif?width=400 400w, public/images/imgtaggen-800.avif?width=800 800w" />
  <source
    type="image/webp"
    srcset="public/images/imgtaggen-100.webp?width=100 100w, public/images/imgtaggen-200.webp?width=200 200w, public/images/imgtaggen-400.webp?width=400 400w, public/images/imgtaggen-800.webp?width=800 800w" />
  <img
    src="public/images/imgtaggen.png"
    srcset="public/images/imgtaggen-100.png?width=100 100w, public/images/imgtaggen-200.png?width=200 200w, public/images/imgtaggen-400.png?width=400 400w, public/images/imgtaggen-800.png?width=800 800w"
    sizes="(max-width: 800px) 100vw, 50vw"
    style="width: 100%; aspect-ratio: 1.791044776119403"
    loading="lazy"
    decoding="async"
    alt="My awesome image"
  />
</picture>
```

## Use this with bimgc

[bimgc](https://bimgc.codewithshin.com/) is a CLI tool for converting PNG and JPG images to AVIF and WebP format with various sizes and saves them in a specified output directory. The output images are named based on the input file and include information about their size and format.

Use `bimgc` in conjunction with `imgtaggen`.

## Reference

- Read more about [Optimal Images in HTML](https://www.builder.io/blog/fast-images)
- Find out more about [Batch Image Converter: bimgc](https://bimgc.codewithshin.com/)

## PWA: Fast & Offline

The [docs website](https://imgtaggen.codewithshin.com/) can be downloaded and installed on your device for offline access as a Progressive Web App.

To install a PWA, look for the "Add to Home Screen" option in the browser's menu or settings. On most mobile devices, this option can be found by visiting the website, then selecting the "Options" or "Menu" button in the browser, and looking for the "Add to Home Screen" option. On some desktop browsers, right-click on the page and select "Install".



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.6.10](https://github.com/shinokada/imgtaggen/compare/v1.6.9...v1.6.10) (2023-02-08)

### [1.6.9](https://github.com/shinokada/imgtaggen/compare/v1.6.8...v1.6.9) (2023-02-06)


### Bug Fixes

* optimized png ([7c5b6d2](https://github.com/shinokada/imgtaggen/commit/7c5b6d298810cb8b4246395d5a79da05a8ff74d8))

### [1.6.8](https://github.com/shinokada/imgtaggen/compare/v1.6.7...v1.6.8) (2023-02-06)

### [1.6.7](https://github.com/shinokada/imgtaggen/compare/v1.6.6...v1.6.7) (2023-02-06)

### [1.6.6](https://github.com/shinokada/imgtaggen/compare/v1.6.5...v1.6.6) (2023-02-06)


### Bug Fixes

* image updates ([a703c58](https://github.com/shinokada/imgtaggen/commit/a703c58ac37ec4646f0881de7c0b5ccf2761916d))

### [1.6.5](https://github.com/shinokada/imgtaggen/compare/v1.6.4...v1.6.5) (2023-02-06)

### [1.6.4](https://github.com/shinokada/imgtaggen/compare/v1.6.3...v1.6.4) (2023-02-05)


### Bug Fixes

* add sizes after image names. update README ([ff5a5ee](https://github.com/shinokada/imgtaggen/commit/ff5a5eecb594b9be5c2d9df3dedc17de1be59517))

### [1.6.3](https://github.com/shinokada/imgtaggen/compare/v1.6.2...v1.6.3) (2023-02-05)

### [1.6.2](https://github.com/shinokada/imgtaggen/compare/v1.6.1...v1.6.2) (2023-02-05)


### Bug Fixes

* update service-worker version ([8990c3c](https://github.com/shinokada/imgtaggen/commit/8990c3c28f87ecbf0b93a8e83811c40b548a1214))

### [1.6.1](https://github.com/shinokada/imgtaggen/compare/v1.6.0...v1.6.1) (2023-02-05)


### Bug Fixes

* add emojis ([989ad61](https://github.com/shinokada/imgtaggen/commit/989ad6104075ef4127f6d1ec801c12bf3fe57b37))

## [1.6.0](https://github.com/shinokada/imgtaggen/compare/v1.5.1...v1.6.0) (2023-02-05)


### Features

* add sizes option ([4c36d04](https://github.com/shinokada/imgtaggen/commit/4c36d0487740e39549e2d57efcc3fabe73d671d4))


### Bug Fixes

* remove top image ([d426ec7](https://github.com/shinokada/imgtaggen/commit/d426ec7d53946c8f5d140db3993821694dc7ba40))

### [1.5.1](https://github.com/shinokada/imgtaggen/compare/v1.5.0...v1.5.1) (2023-02-04)


### Bug Fixes

* change static to public dir ([0e20a10](https://github.com/shinokada/imgtaggen/commit/0e20a1044d95a6f95e765b2617dbc985355ea777))

## [1.5.0](https://github.com/shinokada/imgtaggen/compare/v1.4.0...v1.5.0) (2023-02-04)


### Features

* add positional path/to/image ([8dbabc6](https://github.com/shinokada/imgtaggen/commit/8dbabc68bb1179b0f44afefc7420ee9550ff08e1))

## [1.4.0](https://github.com/shinokada/imgtaggen/compare/v1.3.5...v1.4.0) (2023-02-04)


### Features

* change to commonjs, change node-clipboardy ([ec76c23](https://github.com/shinokada/imgtaggen/commit/ec76c234cf9a653970a638f00076adada18aa7b7))

### [1.3.5](https://github.com/shinokada/imgtaggen/compare/v1.3.4...v1.3.5) (2023-02-04)

### [1.3.4](https://github.com/shinokada/imgtaggen/compare/v1.3.3...v1.3.4) (2023-02-04)


### Bug Fixes

* main in package.json ([3d09acb](https://github.com/shinokada/imgtaggen/commit/3d09acb475d24e87c0d5a4f5ec60755b828d6759))

### [1.3.3](https://github.com/shinokada/imgtaggen/compare/v1.3.2...v1.3.3) (2023-02-04)


### Bug Fixes

* version from package.json ([9d8022d](https://github.com/shinokada/imgtaggen/commit/9d8022dc610e5959a481203bcfe221e9ac05cb10))
* version from package.json ([fc07430](https://github.com/shinokada/imgtaggen/commit/fc07430ed22c4639c9e3862d253a414573ddde18))

### [1.3.2](https://github.com/shinokada/imgtaggen/compare/v1.3.1...v1.3.2) (2023-02-04)

### [1.3.1](https://github.com/shinokada/imgtaggen/compare/v1.3.0...v1.3.1) (2023-02-04)


### Bug Fixes

* index.html ([ac9bf42](https://github.com/shinokada/imgtaggen/commit/ac9bf42c8fd181fbf59ffc3a73ac4294c04bdc7f))

## [1.3.0](https://github.com/shinokada/imgtaggen/compare/v1.2.5...v1.3.0) (2023-02-04)


### Features

* add pwa ([dba4b21](https://github.com/shinokada/imgtaggen/commit/dba4b21ef9e6e9826ae5f4fd14c6cdf715ebcf1e))

### [1.2.5](https://github.com/shinokada/imgtaggen/compare/v1.2.3...v1.2.5) (2023-02-04)

### [1.2.3](https://github.com/shinokada/imgtaggen/compare/v1.2.2...v1.2.3) (2023-02-04)

### [1.2.2](https://github.com/shinokada/imgtaggen/compare/v1.2.1...v1.2.2) (2023-02-04)

### [1.2.1](https://github.com/shinokada/imgtaggen/compare/v1.2.0...v1.2.1) (2023-02-04)

## [1.2.0](https://github.com/shinokada/imgtaggen/compare/v1.1.0...v1.2.0) (2023-02-04)


### Features

* add package script ([4904b41](https://github.com/shinokada/imgtaggen/commit/4904b41066339b217c357bb697fd97bdd6183d40))

## [1.1.0](https://github.com/shinokada/imgtaggen/compare/v1.0.1...v1.1.0) (2023-02-04)


### Features

* add bin dir ([b45dba7](https://github.com/shinokada/imgtaggen/commit/b45dba723d2d2319c34db61eff5f94f5eb17bdd7))

### 1.0.1 (2023-02-04)


service-worker.js
// Change this to your repository name
const GHPATH = 'https://imgtaggen.codewithshin.com/';

// Choose a different app prefix name
const APP_PREFIX = 'imgtaggen_';

// The version of the cache. Every time you change any of the files
// you need to change this version (version_01, version_02…).
// If you don't change the version, the service worker will give your
// users the old files!
const VERSION = 'version_04';

// The files to make available for offline use. make sure to add
// others to this list
const URLS = [
  `${GHPATH}/`,
  `${GHPATH}/index.html`,
  '//cdn.jsdelivr.net/npm/docsify/lib/themes/vue.css',
  '//cdn.jsdelivr.net/npm/docsify@4'
  // `${GHPATH}/js/app.js`
];

const CACHE_NAME = APP_PREFIX + VERSION;
self.addEventListener('fetch', function (e) {
  console.log('Fetch request : ' + e.request.url);
  e.respondWith(
    caches.match(e.request).then(function (request) {
      if (request) {
        console.log('Responding with cache : ' + e.request.url);
        return request;
      } else {
        console.log('File is not cached, fetching : ' + e.request.url);
        return fetch(e.request);
      }
    })
  );
});

self.addEventListener('install', function (e) {
  e.waitUntil(
    caches.open(CACHE_NAME).then(function (cache) {
      console.log('Installing cache : ' + CACHE_NAME);
      return cache.addAll(URLS);
    })
  );
});

self.addEventListener('activate', function (e) {
  e.waitUntil(
    caches.keys().then(function (keyList) {
      var cacheWhitelist = keyList.filter(function (key) {
        return key.indexOf(APP_PREFIX);
      });
      cacheWhitelist.push(CACHE_NAME);
      return Promise.all(
        keyList.map(function (key, i) {
          if (cacheWhitelist.indexOf(key) === -1) {
            console.log('Deleting cache : ' + keyList[i]);
            return caches.delete(keyList[i]);
          }
        })
      );
    })
  );
});

imgtaggen.js
#!/usr/bin/env node

import yargs from "yargs";
import clipboardy from "clipboardy";

yargs(process.argv.slice(2))
  .scriptName("imgtaggen")
  .usage('$0 <cmd> [args]')
  .options({
    noavif: {
      type: 'boolean',
      default: false,
      describe: 'Do not generate an AVIF image tag'
    },
    nowebp: {
      type: 'boolean',
      default: false,
      describe: 'Do not generate a WEBP image tag'
    },
    noclipboard: {
      type: 'boolean',
      default: false,
      describe: 'Do not copy the image tag to the clipboard'
    }
  }).command('*', "Fallback command", (yargs) => {
  }, (argv) => {
    console.log('Generating image tag...')
    console.log('noavif:', argv.noavif)
    console.log('nowebp:', argv.nowebp)
    console.log('noclipboard:', argv.noclipboard)
    generateImgTag(argv.noavif, argv.nowebp, argv.noclipboard);
  })
  .version("1.0.0")
  .help()
  .argv


function generateImgTag (noavif, nowebp, noclipboard) {
  let tag = `<picture>\n`;

  if (!noavif) {
    tag += `  <source\n`;
    tag += `    type="image/avif"\n`;
    tag += `    srcset="/image.avif?width=100 100w, /image.avif?width=200 200w, /image.avif?width=400 400w, /image.avif?width=800 800w" />\n`;
  }

  if (!nowebp) {
    tag += `  <source\n`;
    tag += `    type="image/webp"\n`;
    tag += `    srcset="/image.webp?width=100 100w, /image.webp?width=200 200w, /image.webp?width=400 400w, /image.webp?width=800 800w" />\n`;
  }

  tag += `  <img\n`;
  tag += `    src="/image.png"\n`;
  tag += `    srcset="/image.png?width=100 100w, /image.png?width=200 200w, /image.png?width=400 400w, /image.png?width=800 800w"\n`;
  tag += `    sizes="(max-width: 800px) 100vw, 50vw"\n`;
  tag += `    style="width: 100%; aspect-ratio: 16/9"\n`;
  tag += `    loading="lazy"\n`;
  tag += `    decoding="async"\n`;
  tag += `    alt="Builder.io drag and drop interface"\n`;
  tag += `  />\n`;
  tag += `</picture>\n`;

  

  if (!noclipboard) {
    clipboardy.writeSync(tag);
    console.log("Copied output to the clipboard");
  } else {
    console.log(tag);
  }
}
